internal/services/firewall/firewall_policy_application_rule_collection_resource.go

```diff
@@ -0,0 +1,432 @@
+	}
+	if err := future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		return fmt.Errorf("waiting Firewall Policy Rule Collection Group %q (Resource Group %q / Policy: %q / Rule Collection Group: %q): %+v", id.RuleCollectionName, id.ResourceGroup, id.FirewallPolicyName, id.RuleCollectionGroupName, err)
+	}
+
+	return resourceFirewallPolicyApplicationRuleCollectionRead(d, meta)
+}
+
+func resourceFirewallPolicyApplicationRuleCollectionDelete(d *pluginsdk.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Firewall.FirewallPolicyRuleGroupClient
+	ctx, cancel := timeouts.ForDelete(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := parse.FirewallPolicyRuleCollectionID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	locks.ByName(id.FirewallPolicyName, AzureFirewallPolicyResourceName)
+	defer locks.UnlockByName(id.FirewallPolicyName, AzureFirewallPolicyResourceName)
+
+	resp, err := client.Get(ctx, id.ResourceGroup, id.FirewallPolicyName, id.RuleCollectionGroupName)
+	if err != nil {
+		if !utils.ResponseWasNotFound(resp.Response) {
+			return fmt.Errorf("checking for existing Firewall Policy Rule Collection Group %q (Resource Group %q / Policy %q): %+v", id.RuleCollectionGroupName, id.ResourceGroup, id.FirewallPolicyName, err)
+		}
+	}
+
+	idx := indexFunc(*resp.RuleCollections, func(policy network.BasicFirewallPolicyRuleCollection) bool {
+		info, _ := policy.AsFirewallPolicyFilterRuleCollection()
+		return *info.Name == id.RuleCollectionName
+	})
+
+	if idx == -1 {
+		log.Printf("[DEBUG] Firewall Policy Rule Collection %q was not found in Firewall Policy Rule Collection Group %q - removing from state!", id.RuleCollectionName, id.RuleCollectionGroupName)
+		return nil
+	}
+
+	param := network.FirewallPolicyRuleCollectionGroup{
+		FirewallPolicyRuleCollectionGroupProperties: &network.FirewallPolicyRuleCollectionGroupProperties{
+			Priority: resp.Priority,
+		},
+	}
+	rulesCollections := *resp.RuleCollections
+	rulesCollections = append(rulesCollections[:idx], rulesCollections[idx+1:]...)
+	param.FirewallPolicyRuleCollectionGroupProperties.RuleCollections = &rulesCollections
+
+	future, err := client.CreateOrUpdate(ctx, id.ResourceGroup, id.FirewallPolicyName, id.RuleCollectionGroupName, param)
+	if err != nil {
+		return fmt.Errorf("deleting Firewall Policy Rule Collection %q (Resource Group %q / Policy: %q / Rule Collection Group: %q): %+v", id.RuleCollectionName, id.ResourceGroup, id.FirewallPolicyName, id.RuleCollectionGroupName, err)
+	}
+	if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		if !response.WasNotFound(future.Response()) {
+			return fmt.Errorf("waiting for deleting %q (Resource Group %q / Policy: %q / Rule Collection Group: %q): %+v", id.RuleCollectionName, id.ResourceGroup, id.FirewallPolicyName, id.RuleCollectionGroupName, err)
+		}
+	}
+
+	return nil
+}
+
+func indexFunc[T any](s []T, f func(T) bool) int {
```

Comment:
This probably can be replaced by https://pkg.go.dev/slices#IndexFunc once we are using Go v1.21. So would you mind adding a TODO to mention the replacement?